### PR TITLE
Comments: Add pingbacks and trackbacks to the comments tree

### PR DESCRIPTION
**Pull request by Copons**
Add pingbacks and trackbacks to the `comments-tree` state.

New tree items will have the following properties:
```
{
  commentId,
  commentParentId,
  postId,
  status,
  type: "comment" | "pingback" | "trackback",
}
```

### CODE REVIEW

**Comment by kwight** [comment]
After discussion in p1504797804000109-slack-lannister, this looks good and works great for me 👍 

Tested by firing up the PR, loaded `calypso-localhost:3000/comments/(site-with-pingbacks)`, and verified that the pingbacks were present in `state.comments.trees.(site ID)` (conveniently, any pingbacks or trackbacks will be the last items).

File: client/state/data-layer/wpcom/sites/comments-tree/index.js [hunk 1]
```diff
@@ -35,16 +35,22 @@ export const fetchCommentsTreeForSite = ( { dispatch }, action ) => {
 	);
 };
 
+const mapTree = ( tree, status, type ) => map( tree, comment => ( {
+	commentId: comment[ 0 ],
+	commentParentId: comment[ 2 ],
+	postId: comment[ 1 ],
+	status,
+	type,
+} ) );
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/comments-tree/index.js]: how about using some destructuring to remove the numerical indexing?

```js
map( tree, ( [ commentId, postId, commentParentId ] ) => {
	commentId,
	commentParentId,
	postId,
	status,
	type,
} )
```

File: client/state/data-layer/wpcom/sites/comments-tree/index.js [hunk 2]
```diff
@@ -35,16 +35,22 @@ export const fetchCommentsTreeForSite = ( { dispatch }, action ) => {
...
-		commentId: comment[ 0 ],
-		postId: comment[ 1 ],
-		commentParentId: comment[ 2 ],
-		status,
-	} ) );
+	const tree = [
+		...mapTree( data.comments_tree, status, 'comment' ),
+		...mapTree( data.pingbacks_tree, status, 'pingback' ),
+		...mapTree( data.trackbacks_tree, status, 'trackback' ),
+	];
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/comments-tree/index.js]: this is probably best as-is, but we could also consider another approach…

```js
const tree = [ 'comment', 'pingback', 'trackback' ]
	.map( type => mapTree( data[ `${ type }s_tree` ], status, type ) )
	.reduce( ( final, next ) => final.concat( next ), [] ); // could also be something like ( a, b ) => [ ...a, ...b ]
```

File: client/state/data-layer/wpcom/sites/comments-tree/index.js (hunk 1 above)
Comment (Copons) [reply client/state/data-layer/wpcom/sites/comments-tree/index.js]: Oooh I like this!

File: client/state/data-layer/wpcom/sites/comments-tree/index.js (hunk 2 above)
Comment (Copons) [reply client/state/data-layer/wpcom/sites/comments-tree/index.js]: You know I've like spent half the day thinking of a pretty functional way to do this (`ary`, `partial`, `flow`, `rearg`, and whatnot), and in the end settled with the simplest approach I could write. 🙂 
With your approach we could drop the `mapTree()` utility and keep all contained in a single expression - which was what I was aiming for anyway!

Comment (dmsnell) [reply client/state/data-layer/wpcom/sites/comments-tree/index.js]: Functional shouldn't mean hard to follow; if it does it's being forced and not applied appropriately.

I see now what's going on: we're wanting to run each item through the shape-converter then join them all into a single list.

We have two typical functions to help us here: `map` and `flatMap` (which runs each item through a collection then flattens the results). In this case it's a little odd just because of the property accesses involved. We could turn this entire thing though into something like this…

```js
flatMap(
	[ 'comment', 'pingback', 'trackback' ],
	type => map(
		data[ `${ type }s_tree` ],
		( [ commentId, … ] ) => ( { … } )
	)
);
```

…but at this point although it's maybe slightly faster I don't find it clearer. Your construction is clear.

Comment (vindl) [reply client/state/data-layer/wpcom/sites/comments-tree/index.js]: I like the `flatMap` solution, but the original one, even though it's somewhat repetitive, still looks more readable to me. Since it's unlikely that we'll be adding more types to the tree anytime soon, I'd say stick with the original approach.

Comment (Copons) [reply client/state/data-layer/wpcom/sites/comments-tree/index.js]: We might actually add Reblogs at some point, but I've no idea of how they're currently treated and stored. 😄
Though, it would be super easy to add it to the current structure or replace it with the `flatMap`.
